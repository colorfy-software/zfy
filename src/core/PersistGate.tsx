import React, { useEffect } from 'react'

import type { CreateStoreType } from '../types'

import useRehydrate from './use-rehydrate'
import { validatePersistGate } from '../internals/validation'

interface PropsType<StoresType> {
  children: JSX.Element | JSX.Element[] | (() => JSX.Element)
  loader?: JSX.Element | (() => JSX.Element)
  stores: StoresType
}

/**
 * Component that renders a loader while all the provided persisted stores get rehydrated.
 * @prop stores - `Record<string, CreateStoreType<any>>`— Object containing all the zustand stores to rehydrate.
 * @prop children - `JSX.Element | JSX.Element[]`— React element to render once rehydration is completed.
 * @prop loader - `JSX.Element`— Optional. Loader component to display.
 */
const PersistGate = <StoresType extends Record<string, CreateStoreType<any>>>({
  stores,
  loader,
  children,
}: PropsType<StoresType>): JSX.Element | null => {
  useEffect(() => {
    validatePersistGate(stores)
  }, [stores])

  const isRehydrated = useRehydrate<StoresType>(stores)

  return (
    <>
      {isRehydrated
        ? typeof children === 'function'
          ? children()
          : children
        : typeof loader === 'function'
        ? loader()
        : loader || null}
    </>
  )
}

export default PersistGate
