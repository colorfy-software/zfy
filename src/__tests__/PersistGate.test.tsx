import React from 'react'
import { Text, View } from 'react-native'
import { render, waitFor } from '@testing-library/react-native'

import { data, sleep, SyncStorage, AsyncStorage, rehydratedData } from '.'
import PersistGate from '../core/PersistGate'
import createStore from '../core/create-store'

describe('ðŸšª Core > PersistGate:', () => {
  it('renders loader and rehydrates with sync storage', async () => {
    const loader = jest.fn()
    const store = createStore('jest', data, {
      persist: { getStorage: () => SyncStorage },
    })

    const { toJSON } = await waitFor(() =>
      render(
        <PersistGate stores={[store]} loader={loader}>
          <View>
            <Text>My App</Text>
          </View>
        </PersistGate>
      )
    )

    expect(toJSON()).toMatchSnapshot()
    expect(loader).toHaveBeenCalled()
    expect(store.getState().data).toEqual(rehydratedData)

    expect.assertions(3)
  })

  it('renders loader and rehydrates with async storage', async () => {
    const loader = jest.fn()
    const store = createStore('jest', data, {
      persist: { getStorage: () => AsyncStorage },
    })

    // FIXME: Fix act issue:
    // Warning: An update to PersistGate inside a test was not wrapped in act(...).
    const { toJSON } = render(
      <PersistGate stores={[store]} loader={loader}>
        <View>
          <Text>My App</Text>
        </View>
      </PersistGate>
    )

    expect(toJSON()).toMatchSnapshot()
    expect(loader).toHaveBeenCalled()
    await sleep(250)

    expect(store.getState().data).toEqual(rehydratedData)

    expect.assertions(3)
  })
})
