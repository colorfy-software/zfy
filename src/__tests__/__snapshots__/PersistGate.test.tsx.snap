// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ğŸšª Core > PersistGate: renders loader and rehydrates with async storage 1`] = `null`;

exports[`ğŸšª Core > PersistGate: renders loader and rehydrates with sync storage 1`] = `
<View>
  <Text>
    My App
  </Text>
</View>
`;
